text/src/autogluon/text/automm/models/categorical_transformer.py

```diff
@@ -0,0 +1,275 @@
+
+        for parameter in [self.embeddings.weight, self.bias]:
+            if parameter is not None:
+                initialization_.apply(parameter, d_token)
+
+    @property
+    def n_tokens(self) -> int:
+        """The number of tokens."""
+        return len(self.category_offsets)
+
+    @property
+    def d_token(self) -> int:
+        """The size of one token."""
+        return self.embeddings.embedding_dim
+
+    def forward(
+            self,
+            x: Tensor,
+    ) -> Tensor:
+        
+        x = self.embeddings(x + self.category_offsets[None])
+
+        if self.bias is not None:
+            x = x + self.bias[None]
+
+        return x
+
+
+
+
+class  CategoricalTransformer(nn.Module):
+    """The FT-Transformer model proposed in [gorishniy2021revisiting].
+
+    Transforms features to tokens with `FeatureTokenizer` and applies `Transformer` [vaswani2017attention]
+    to the tokens. The following illustration provides a high-level overview of the
+    architecture:
+
+    .. image:: ../images/ft_transformer.png
+        :scale: 25%
+        :alt: FT-Transformer
+
+    The following illustration demonstrates one Transformer block for :code:`prenormalization=True`:
+
+    .. image:: ../images/transformer_block.png
+        :scale: 25%
+        :alt: PreNorm Transformer block
+
+    Examples:
+        .. testcode::
+
+            x_num = torch.randn(4, 3)
+            x_cat = torch.tensor([[0, 1], [1, 0], [0, 2], [1, 1]])
+
+            module = FTTransformer.make_baseline(
+                n_num_features=3,
+                cat_cardinalities=[2, 3],
+                d_token=8,
+                n_blocks=2,
+                attention_dropout=0.2,
+                ffn_d_hidden=6,
+                ffn_dropout=0.2,
+                residual_dropout=0.0,
+                d_out=1,
+            )
+            x = module(x_num, x_cat)
+            assert x.shape == (4, 1)
+
+            module = FTTransformer.make_default(
+                n_num_features=3,
+                cat_cardinalities=[2, 3],
+                d_out=1,
+            )
+            x = module(x_num, x_cat)
+            assert x.shape == (4, 1)
+
+        To learn more about the baseline and default parameters:
+
+        .. testcode::
+
+            baseline_parameters = FTTransformer.get_baseline_transformer_subconfig()
+            default_parameters = FTTransformer.get_default_transformer_config()
+
+    References:
+        * [gorishniy2021revisiting] Yury Gorishniy, Ivan Rubachev, Valentin Khrulkov, Artem Babenko, "Revisiting Deep Learning Models for Tabular Data", 2021
+        * [vaswani2017attention] Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit, Llion Jones, Aidan N. Gomez, Lukasz Kaiser, Illia Polosukhin, "Attention Is All You Need", 2017
+    """
+
+    def __init__(
+        self, 
+        prefix: str, 
+        num_categories: List[int],
+        d_token: int,
+        # n_tokens: Optional[int] = None,
```

Comment:
Consider to remove the comment.

Reply:
Unrelated comments are properly removed.